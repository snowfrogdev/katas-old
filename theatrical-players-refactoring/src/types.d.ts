interface Invoice {
  customer: string;
  performances: Performance[];
}

interface Performance {
  playID: string;
  audience: number;
}

interface Statement {
  customer: string;
  performances: Array<{ playName: string; audience: number; amount: string }>;
  totalAmount: string;
  volumeCredits: number;
}

interface Play {
  name: string;
  type: string;
}

interface PrintStrategy {
  print(invoice: Invoice): string
}