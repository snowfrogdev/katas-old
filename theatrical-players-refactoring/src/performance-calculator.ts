export abstract class PerformanceCalculator {
  constructor(protected performance: Performance) {}
  abstract calculateAmount(): number;

  public calculateVolumeCredits(): number {
    let volumeCredits = 0;
    volumeCredits += Math.max(this.performance.audience - 30, 0);
    return volumeCredits;
  }
}

export class ComedyPerformanceCalculator extends PerformanceCalculator {
  constructor(performance: Performance) {
    super(performance);
  }
  calculateAmount(): number {
    let amount = 300_00;
    if (this.performance.audience > 20) {
      amount += 100_00 + 5_00 * (this.performance.audience - 20);
    }
    amount += 3_00 * this.performance.audience;
    return amount;
  }

  calculateVolumeCredits() {
    return super.calculateVolumeCredits() + Math.floor(this.performance.audience / 5);
  }
}

export class TragedyPerformanceCalculator extends PerformanceCalculator {
  constructor(performance: Performance) {
    super(performance);
  }
  calculateAmount(): number {
    let amount = 400_00;
    if (this.performance.audience > 30) {
      amount += 10_00 * (this.performance.audience - 30);
    }
    return amount;
  }
}

export class MysteryPerformanceCalculator extends PerformanceCalculator {
  constructor(performance: Performance) {
    super(performance);
  }
  calculateAmount(): number {
    let amount = 250_00;
    if (this.performance.audience > 15) {
      amount += 15_00 * (this.performance.audience - 15);
    }
    return amount;
  }
}

export class HorrorPerformanceCalculator extends PerformanceCalculator {
  constructor(performance: Performance) {
    super(performance);
  }
  calculateAmount(): number {
    let amount = 600_00;
    if (this.performance.audience > 20) {
      amount += 5_00 * (this.performance.audience - 20);
    }
    return amount;
  }
}
