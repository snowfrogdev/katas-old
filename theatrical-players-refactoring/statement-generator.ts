enum PlayType {
  Tragedy = 'tragedy',
  Comedy = 'comedy'
}

export class StatementGenerator {
  constructor(private plays: Map<string, Play>) {}

  generateStatement(invoice: Invoice): Statement {
    return {
      customer: invoice.customer,
      performances: invoice.performances.map(performance => {
        const performanceCalculator = this.createPerformanceCalculator(performance);
        return {
          audience: performance.audience,
          playName: this.plays.get(performance.playID)!.name,
          amount: this.formatCentsToUSD(
            performanceCalculator.calculatePerformanceAmount(performance)
          )
        };
      }),
      totalAmount: this.formatCentsToUSD(this.calculateTotalAmount(invoice)),
      volumeCredits: this.calculateTotalVolumeCredits(invoice)
    };
  }

  private formatCentsToUSD(value: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2
    }).format(value / 100);
  }

  public calculateTotalAmount(invoice: Invoice): number {
    return invoice.performances.reduce((total, performance) => {
      const performanceCalculator = this.createPerformanceCalculator(performance);
      return total + performanceCalculator.calculatePerformanceAmount(performance);
    }, 0);
  }

  public calculateTotalVolumeCredits(invoice: Invoice): number {
    return invoice.performances.reduce((total, performance) => {
      const performanceCalculator = this.createPerformanceCalculator(performance);
      return total + performanceCalculator.calculatePerformanceVolumeCredits(performance);
    }, 0);
  }

  private createPerformanceCalculator(performance: Performance): PerformanceCalculator {
    const playType = this.plays.get(performance.playID)!.type;
    switch (playType) {
      case PlayType.Comedy:
        return new ComedyPerformanceCalculator(this.plays);
      case PlayType.Tragedy:
        return new PerformanceCalculator(this.plays);
      default:
        throw new Error(`unknown type: ${playType}`);
    }
  }
}

export class PerformanceCalculator {
  constructor(private plays: Map<string, Play>) {}

  public calculatePerformanceAmount(performance: Performance): number {
    let amount = 0;
    const playType = this.plays.get(performance.playID)!.type;
    switch (playType) {
      case PlayType.Tragedy:
        amount = 400_00;
        if (performance.audience > 30) {
          amount += 10_00 * (performance.audience - 30);
        }
        break;
      case PlayType.Comedy:
        throw 'Should be implemented in subclass';
      default:
        throw new Error(`unknown type: ${playType}`);
    }
    return amount;
  }

  public calculatePerformanceVolumeCredits(performance: Performance): number {
    let volumeCredits = 0;
    volumeCredits += Math.max(performance.audience - 30, 0);
    if (PlayType.Comedy === this.plays.get(performance.playID)!.type)
      volumeCredits += Math.floor(performance.audience / 5);
    return volumeCredits;
  }
}

class ComedyPerformanceCalculator extends PerformanceCalculator {
  calculatePerformanceAmount(performance: Performance): number {
    let amount = 300_00;
    if (performance.audience > 20) {
      amount += 100_00 + 5_00 * (performance.audience - 20);
    }
    amount += 3_00 * performance.audience;
    return amount;
  }
}
