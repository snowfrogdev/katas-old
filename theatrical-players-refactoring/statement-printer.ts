interface Play {
  name: string;
  type: string;
}

enum PlayType {
  Tragedy = 'tragedy',
  Comedy = 'comedy'
}

interface Performance {
  playID: string;
  audience: number;
}

interface Invoice {
  customer: string;
  performances: Performance[];
}

interface Statement {
  customer: string;
  performances: Array<{ playName: string; audience: number; amount: string }>;
  totalAmount: string;
  volumeCredits: number;
}

export class PerformanceCalculator {
  constructor(private plays: Map<string, Play>) {}

  public calculatePerformanceAmount(performance: Performance): number {
    let amount = 0;
    const playType = this.plays.get(performance.playID)!.type;
    switch (playType) {
      case PlayType.Tragedy:
        amount = 400_00;
        if (performance.audience > 30) {
          amount += 10_00 * (performance.audience - 30);
        }
        break;
      case PlayType.Comedy:
        amount = 300_00;
        if (performance.audience > 20) {
          amount += 100_00 + 5_00 * (performance.audience - 20);
        }
        amount += 3_00 * performance.audience;
        break;
      default:
        throw new Error(`unknown type: ${playType}`);
    }
    return amount;
  }

  public calculatePerformanceVolumeCredits(performance: Performance): number {
    let volumeCredits = 0;
    volumeCredits += Math.max(performance.audience - 30, 0);
    if (PlayType.Comedy === this.plays.get(performance.playID)!.type)
      volumeCredits += Math.floor(performance.audience / 5);
    return volumeCredits;
  }
}

export class InvoiceCalculator {
  constructor(private performanceCalculator: PerformanceCalculator) {}
  public calculateTotalAmount(invoice: Invoice): number {
    return invoice.performances.reduce(
      (total, performance) =>
        total + this.performanceCalculator.calculatePerformanceAmount(performance),
      0
    );
  }

  public calculateTotalVolumeCredits(invoice: Invoice): number {
    return invoice.performances.reduce(
      (total, performance) =>
        total + this.performanceCalculator.calculatePerformanceVolumeCredits(performance),
      0
    );
  }
}

export class StatementPrinter {
  constructor(
    private plays: Map<string, Play>,
    private performanceCalculator: PerformanceCalculator,
    private invoiceCalculator: InvoiceCalculator
  ) {}
  print(invoice: Invoice): string {
    const statement = this.generateStatement(invoice);
    return this.produceStatementText(statement);
  }

  private produceStatementText(statement: Statement): string {
    let result = `Statement for ${statement.customer}\n`;
    for (const performance of statement.performances) {
      // print line for this order
      result += ` ${performance.playName}: ${performance.amount} (${
        performance.audience
      } seats)\n`;
    }
    result += `Amount owed is ${statement.totalAmount}\n`;
    result += `You earned ${statement.volumeCredits} credits\n`;
    return result;
  }

  private generateStatement(invoice: Invoice): Statement {
    return {
      customer: invoice.customer,
      performances: invoice.performances.map(performance => ({
        audience: performance.audience,
        playName: this.plays.get(performance.playID)!.name,
        amount: this.formatCentsToUSD(this.performanceCalculator.calculatePerformanceAmount(performance))
      })),
      totalAmount: this.formatCentsToUSD(this.invoiceCalculator.calculateTotalAmount(invoice)),
      volumeCredits: this.invoiceCalculator.calculateTotalVolumeCredits(invoice)
    };
  }

  private formatCentsToUSD(value: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2
    }).format(value / 100);
  }
}

export function convertObjectToMap<T>(object: { [key: string]: T }): Map<string, T> {
  return new Map(Object.entries(object));
}
