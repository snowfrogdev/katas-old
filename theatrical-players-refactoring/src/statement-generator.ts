import { PerformanceCalculator, ComedyPerformanceCalculator, TragedyPerformanceCalculator } from "./performance-calculator";

export enum PlayType {
  Tragedy = 'tragedy',
  Comedy = 'comedy'
}

export class StatementGenerator {
  constructor(private plays: Map<string, Play>) {}

  generateStatement(invoice: Invoice): Statement {
    return {
      customer: invoice.customer,
      performances: invoice.performances.map(performance => {
        const performanceCalculator = this.createPerformanceCalculator(performance);
        return {
          audience: performance.audience,
          playName: this.plays.get(performance.playID)!.name,
          amount: this.formatCentsToUSD(
            performanceCalculator.calculatePerformanceAmount(performance)
          )
        };
      }),
      totalAmount: this.formatCentsToUSD(this.calculateTotalAmount(invoice)),
      volumeCredits: this.calculateTotalVolumeCredits(invoice)
    };
  }

  private formatCentsToUSD(value: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2
    }).format(value / 100);
  }

  public calculateTotalAmount(invoice: Invoice): number {
    return invoice.performances.reduce((total, performance) => {
      const performanceCalculator = this.createPerformanceCalculator(performance);
      return total + performanceCalculator.calculatePerformanceAmount(performance);
    }, 0);
  }

  public calculateTotalVolumeCredits(invoice: Invoice): number {
    return invoice.performances.reduce((total, performance) => {
      const performanceCalculator = this.createPerformanceCalculator(performance);
      return total + performanceCalculator.calculatePerformanceVolumeCredits(performance);
    }, 0);
  }

  private createPerformanceCalculator(performance: Performance): PerformanceCalculator {
    const playType = this.plays.get(performance.playID)!.type;
    switch (playType) {
      case PlayType.Comedy:
        return new ComedyPerformanceCalculator(this.plays);
      case PlayType.Tragedy:
        return new TragedyPerformanceCalculator(this.plays);
      default:
        throw new Error(`unknown type: ${playType}`);
    }
  }
}
