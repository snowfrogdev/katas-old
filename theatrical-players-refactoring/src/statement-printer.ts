import { StatementGenerator } from './statement-generator';

export class StatementPrinter {
  constructor(private printStrategy: PrintStrategy) {}
  print(invoice: Invoice): string {
    return this.printStrategy.print(invoice);
  }

  setPrintStrategy(printStrategy: PrintStrategy) {
    this.printStrategy = printStrategy;
  }
}

export class PrintHtmlStrategy implements PrintStrategy {
  constructor(private generator: StatementGenerator) {}
  print(invoice: Invoice): string {
    const statement = this.generator.generateStatement(invoice);
    return this.produceStatementHtml(statement);
  }

  private produceStatementHtml(statement: Statement): string {
    let result = `<h1>Statement for ${statement.customer}</h1>\n`;
    for (const performance of statement.performances) {
      // print line for this order
      result += ` <p>${performance.playName}: ${performance.amount} (${performance.audience} seats)</p>\n`;
    }
    result += `<p>Amount owed is <strong>${statement.totalAmount}</strong></p>\n`;
    result += `<p>You earned <strong>${statement.volumeCredits}</strong> credits</p>\n`;
    return result;
  }
}

export class PrintTextStrategy implements PrintStrategy {
  constructor(private generator: StatementGenerator) {}
  print(invoice: Invoice): string {
    const statement = this.generator.generateStatement(invoice);
    return this.produceStatementText(statement);
  }

  private produceStatementText(statement: Statement): string {
    let result = `Statement for ${statement.customer}\n`;
    for (const performance of statement.performances) {
      // print line for this order
      result += ` ${performance.playName}: ${performance.amount} (${performance.audience} seats)\n`;
    }
    result += `Amount owed is ${statement.totalAmount}\n`;
    result += `You earned ${statement.volumeCredits} credits\n`;
    return result;
  }
}

export function convertObjectToMap<T>(object: { [key: string]: T }): Map<string, T> {
  return new Map(Object.entries(object));
}
