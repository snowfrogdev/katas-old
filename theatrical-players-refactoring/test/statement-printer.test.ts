import { StatementPrinter, convertObjectToMap } from '../src/statement-printer';
import { StatementGenerator } from '../src/statement-generator';
import * as invoice from './invoice.json';
import * as plays from './plays.json';
import * as newInvoice from './invoice_new_plays.json';
import * as newPlays from './new_plays.json';

test('example text statement', () => {
  const playsMap = convertObjectToMap(plays);
  const generator = new StatementGenerator(playsMap);
  const printer = new StatementPrinter(generator);
  expect(printer.print(invoice)).toMatchSnapshot();
});

test('example html statement', () => {
  const playsMap = convertObjectToMap(plays);
  const generator = new StatementGenerator(playsMap);
  const printer = new StatementPrinter(generator);
  expect(printer.printHtml(invoice)).toMatchSnapshot();
})

test('statement with new play types', () => {
  expect(() => {
    const newPlaysMap = convertObjectToMap(newPlays);
    const generator = new StatementGenerator(newPlaysMap);
    const printer = new StatementPrinter(generator);
    printer.print(newInvoice);
  }).toThrow(/unknown type/);
});
