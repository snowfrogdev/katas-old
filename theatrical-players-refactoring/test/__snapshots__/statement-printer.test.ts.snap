// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example html statement 1`] = `
"<h1>Statement for BigCo</h1>
 <p>Hamlet: $650.00 (55 seats)</p>
 <p>As You Like It: $580.00 (35 seats)</p>
 <p>Othello: $500.00 (40 seats)</p>
<p>Amount owed is <strong>$1,730.00</strong></p>
<p>You earned <strong>47</strong> credits</p>
"
`;

exports[`example text statement 1`] = `
"Statement for BigCo
 Hamlet: $650.00 (55 seats)
 As You Like It: $580.00 (35 seats)
 Othello: $500.00 (40 seats)
Amount owed is $1,730.00
You earned 47 credits
"
`;
